Add capability registry

dschwen opened the pull request:
Closes #26470

dschwen committed:
Add capability registry in app (#26470)

dschwen committed:
Add capability support in tester (#26470)

loganharbour commented on framework/include/base/MooseApp.h:
```diff
@@ -31,13 +31,16 @@
 #include "libmesh/mesh_base.h"
 #include "libmesh/point.h"
 
+#include "nlohmann/json.h"
+
```
Move to source

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -609,6 +613,74 @@ MooseApp::MooseApp(InputParameters parameters)
                     name(),
                     " to remove this deprecation warning.");
 
+  // register capabilities
```
Can we move this into a method so that we don't bloat the constructor even more?

loganharbour reviewed (commented):
Looks great.

Should be able to get an easy unit test that instantiates an app and tries to parse the output, not caring what's in it

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1977436/) on 330361e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26471/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1977451/) on 330361e wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3833a5</th>
      <th colspan=4>#26471 330361</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.07%</td>
      <td>85.07%</td>
      <td>+0.00%</td>
      <td>92.68%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>98390</td>
      <td>98427</td>
      <td>+37</td>
      <td>38</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17264</td>
      <td>17268</td>
      <td>+4</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26471/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26471/coverage/framework)


## Modules coverage

Mutiple head shas were found when retrieving head reports:
  330361e461ecb601ff978dd8bc52f11fe1082a88
  6dbeda32b800741fad482c21f5dd5f58d3445e34

### Chemical reactions

<table>
  <thead>
    <tr>
      <th></th>
      <th>3833a5</th>
      <th colspan=4>#26471 330361</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>92.55%</td>
      <td>92.88%</td>
      <td>+0.33%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1441</td>
      <td>1447</td>
      <td>+6</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>116</td>
      <td>111</td>
      <td>-5</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26471/coverage_diff/chemical_reactions)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26471/coverage/chemical_reactions)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/combined)
- `contact` (FAILED)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/functional_expansion_tools)
- `geochemistry` (FAILED)
- `heat_transfer` (FAILED)
- `level_set` (FAILED)
- `misc` (FAILED)
- `navier_stokes` (FAILED)
- `optimization` (FAILED)
- `peridynamics` (FAILED)
- `phase_field` (FAILED)
- `porous_flow` (FAILED)
- `ray_tracing` (FAILED)
- `rdg` (FAILED)
- `reactor` (FAILED)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/richards)
- `scalar_transport` (FAILED)
- `solid_properties` (FAILED)
- `stochastic_tools` (FAILED)
- `tensor_mechanics` (FAILED)
- `thermal_hydraulics` (FAILED)
- `xfem` (FAILED)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1977451/) on 330361e wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 33.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 85.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `tensor_mechanics` coverage rate (required: 84.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 89.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

dschwen committed:
Update test specs (#26470)

dschwen committed:
Rework into singleton registry (#26470)

dschwen committed:
Spelling (#26470)

dschwen committed:
Implement capabilities check in C++ (#26470)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1989994/) on c7587cc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format ad409367e365d3227cf9c1a0df63b3b9cde9b916`

dschwen commented:
Rats, on min gcc the converting constructor for `std::variant` puts a `const char *` argument in `bool` rather than `std::string`. I'll need to add a manual overload to `add`.

dschwen committed:
Work around ambiguous conversion (#26470)

dschwen committed:
Namespace, factor out registration, pytest fix (#26470)

dschwen committed:
Executable can check required capabilities (#26470)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1995528/) on 1392415 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format c1cce92fac6099ee3fde58946b9f7ffd24275fea`

GiudGiud commented on test/tests/meshgenerators/file_mesh_generator/tests:
```diff
@@ -30,7 +30,7 @@
     recover = false
     exodus_version = '>=8.0'
```
should be in the capability?

GiudGiud commented on test/tests/meshgenerators/distributed_rectilinear/generator/tests:
```diff
@@ -40,7 +40,7 @@
     input = 'distributed_rectilinear_mesh_generator.i'
     exodiff = 'distributed_rectilinear_mesh_generator_out_3d_ptscotch.e'
     ptscotch = true
```
could be a capability?

GiudGiud reviewed (commented):
reviewed the tests

dschwen commented on modules/combined/test/tests/DiffuseCreep/tests:
```diff
@@ -13,14 +13,12 @@
     input = 'stress.i'
     exodiff = 'stress_out.e'
-    compiler = 'GCC CLANG'
     custom_cmp = 'stress.cmp'
```
not yet

dschwen replied:
I implemented a full fledged logic parser. So this would be 

```
capabilities = 'compiler=gcc | compiler=clang'
```

dschwen committed:
Address comments (#26470)

dschwen commented:
I have an idea for a hybrid check (first stage by the test harness, second stage by the executable). Some capabilities are set by the base application and are built into the executable. If any of those capabilities is not present we do not have to launch the executable for a dynamic check.

In fact this should be easy to implement! Currently a capability that's not in the JSON at all is assumed to be nonexistent. For the test harness check we'll return a status of _unsure_ instead. Then instead of not registering a capability in the main app, when we know that it cannot possibly be fulfilled (e.g. because libmesh was compile without it) we explicitly set it to `false`.

This should allow us to shortcircuit executable runs for the vast majority of capability checks (`ad_size`, `petsc`, `parmetis`, `mumps`, `method`, `compiler` yada yada yada...).

For capabilities we cannot be sure about, because they might be provided by a dynamically linked executable, we just run the app and use its capability check.

dschwen commented:
Things that are not _inate_ to the executable but refer to a mode of running the executable (e.g. `valgrind`, `recover`, `min_parallel`, threading, etc.) will _not_ be part of the capability system.

Lastly, we should also add a `required_capabilities` parameter to the `[Application]` block. That block should get parsed super early, and dynamic libraries loaded, just before we run the in-executable capabilities check. This would permit users to tag inputs with application capability requirements. 

Really lastly, in a follow-on PR (after the application block stuff), I will add documentation strings to each capability. This will allow us to output human readable error messages detailing why the current executable  cannot run the given input, and hints how to remedy this.

dschwen committed:
Explicitly set to false (#26470)

dschwen committed:
Annotate capabilities with doc strings (#26470)

dschwen committed:
Output machine readable data to file (#26555)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2001186/) on d750613 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format aab76d5271d84a92d1d1e0b4e17aa757a1f4a612`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2002471/) on 8d227cd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a4b4331e3af5aa4ef29a557580d21d72b3a90f57`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2004505/) on 5f88570 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e9f5f109e610453316f89158ccedd8f0f4747199`

loganharbour commented on framework/src/base/Capabilities.C:
```diff
@@ -0,0 +1,208 @@
+    {
+      // error
+      if (test_value.empty())
+        return {false, condition + " broken"};
```
what is broken?

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -610,9 +618,370 @@ MooseApp::MooseApp(InputParameters parameters)
+        "method", method, "The executable was built with METHOD=\"" + method + "\"");
+  }
+
+  {
+    const auto doc = "Workbench Analysis Sequence Processors (WASP) parser library";
+#ifdef WASP_ENABLED
+    haveCapability("wasp", doc);
+#else
+    missingCapability("wasp", doc, " Unable to parse input files.");
+#endif
+  }
```
WASP_ENABLED shouldn't exist anymore

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -610,9 +618,370 @@ MooseApp::MooseApp(InputParameters parameters)
+#endif
+  }
+
+  {
+    const std::string version = QUOTE(LIBMESH_DETECTED_PETSC_VERSION_MAJOR) "." QUOTE(
+        LIBMESH_DETECTED_PETSC_VERSION_MINOR) "." QUOTE(LIBMESH_DETECTED_PETSC_VERSION_SUBMINOR);
+    Moose::Capabilities::add("petsc", version, "Using PETSc version " + version + ".");
+  }
```
what if we don't detect a petsc version (non-release hash)

dschwen replied:
Hmm, yeah, what then? We cannot know if a random hash is greater or less than a specific version (at least not without inspecting the PETSc repo...)

dschwen replied:
@loganharbour we should discuss this. I cannot remember the last time we actually tagged a PETSc non-release. If that will be required we could add an additional capability that is set to `false` if no non-release is used, and to the hash (string) if it is, and conversely we could set `petsc` to `false` if a non-release hash is used.

dschwen replied:
Or set `petsc` to `true`. Any comparison with a bool `true` capability succeeds and any comparison with a bool `false` capability fails. That can be used to engineer "always match" versions or "always fail" versions.

dschwen replied:
We'll table the non-release petsc issue for now.

dschwen commented on framework/include/base/Capabilities.h:
```diff
@@ -0,0 +1,68 @@
+ * the AD backing store size), a string (like the compiler used to build the executable),
+ * or a version number (numbers separated by dots, e.g. the petsc version).
+ */
+class Capabilities
```
Is it? https://github.com/idaholab/moose/blob/next/framework/include/base/Registry.h

loganharbour replied:
excuse me, newer registries :)

if this wasn't like this, you could test with fresh capabilities in unit tests

dschwen replied:
You mean like this beautiful construct?
```
moose::internal::ExecFlagRegistry::getExecFlagRegistry().getDefaultFlags();
```

loganharbour replied:
so good

dschwen replied:
Yeah, thats :hankey: ... I need `FRIEND_TEST` to work for that super feature of fresh capabilities in unit tests. I think that means ripping `Capabilities` out of the `Moose` namespace :-(

dschwen replied:
I think I figure it out. Namespaces and friend classes require a proper forward declaration.

loganharbour commented:
> Sorry to continue with my tradition of nit-picking naming choices, but my reaction when I saw this is that Capabilities is a very broad term

The problem is that I'd supposed to be very broad

dschwen committed:
Change to new registry pattern (#26470)

dschwen committed:
GTest is only available in unit tests (#26470)

dschwen commented on framework/src/base/Capabilities.C:
```diff
@@ -0,0 +1,208 @@
+      it_pair->second.first == value)
+    it_pair = _capability_registry.emplace_hint(it_pair, capability, std::make_pair(value, doc));
+  else
+    mooseWarning("Capability '", capability, "' was already registered with a different value.");
```
I'm doing it, but making it print `true`/`false` would be a bit much work...

dschwen replied:
I figured out a neat way with fold expressions...

dschwen committed:
Fix multiple capability check (#26470)

dschwen committed:
min gcc fix (#26470)

loganharbour commented on python/TestHarness/util.py:
```diff
@@ -616,6 +616,112 @@ def checkLibtorchVersion(checks, test):
+        results = json.loads(output)
+    except IndexError:
+        return {}
+    except json.decoder.JSONDecodeError:
+        print(f'{exe} --show-capabilities, produced invalid JSON output')
+        sys.exit(1)
```
```suggestion
    except json.decoder.JSONDecodeError as e:
        raise Exception(f'{exe} --show-capabilities, produced invalid JSON output') from e
```

I've started doing this, much better

dschwen committed:
Exit code processing (#26470)

dschwen committed:
Fix exception test (#26470)

loganharbour reviewed (commented):
Can we talk the `MooseMain` strategy again soon? We had a few other changes that we wanted to make that I wanted to make sure were easily compatible

loganharbour commented:
What's the verdict on this one?

dschwen commented:
I've started working on the python module wrapper.

dschwen committed:
Add python bindings (#26470)

dschwen commented:
I'm perplexed by this error:

```
pycapabilities/tests.capabilities: Working Directory: /data/civet0/build/moose/python/pycapabilities/tests
pycapabilities/tests.capabilities: Running command: python3 -m unittest -v test_capabilities
pycapabilities/tests.capabilities: Building and linking capabilities...
pycapabilities/tests.capabilities: test (test_capabilities.TestCapabilities) ... ERROR
pycapabilities/tests.capabilities: 
pycapabilities/tests.capabilities: ======================================================================
pycapabilities/tests.capabilities: ERROR: test (test_capabilities.TestCapabilities)
pycapabilities/tests.capabilities: ----------------------------------------------------------------------
pycapabilities/tests.capabilities: Traceback (most recent call last):
pycapabilities/tests.capabilities:   File "/data/civet0/build/moose/python/pycapabilities/__init__.py", line 12, in <module>
pycapabilities/tests.capabilities:     import capabilities
pycapabilities/tests.capabilities: ModuleNotFoundError: No module named 'capabilities'
pycapabilities/tests.capabilities: 
pycapabilities/tests.capabilities: During handling of the above exception, another exception occurred:
pycapabilities/tests.capabilities: 
pycapabilities/tests.capabilities: Traceback (most recent call last):
pycapabilities/tests.capabilities:   File "/data/civet0/build/moose/python/pycapabilities/__init__.py", line 17, in <module>
pycapabilities/tests.capabilities:     import capabilities
pycapabilities/tests.capabilities: ImportError: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /data/civet0/build/moose/framework/contrib/capabilities/capabilities.so)
pycapabilities/tests.capabilities: 
pycapabilities/tests.capabilities: During handling of the above exception, another exception occurred:
pycapabilities/tests.capabilities: 
pycapabilities/tests.capabilities: Traceback (most recent call last):
pycapabilities/tests.capabilities:   File "/data/civet0/build/moose/python/pycapabilities/tests/test_capabilities.py", line 18, in test
pycapabilities/tests.capabilities:     import pycapabilities
pycapabilities/tests.capabilities:   File "/data/civet0/build/moose/python/pycapabilities/__init__.py", line 21, in <module>
pycapabilities/tests.capabilities:     import capabilities
pycapabilities/tests.capabilities: ImportError: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /data/civet0/build/moose/framework/contrib/capabilities/capabilities.so)
pycapabilities/tests.capabilities: 
pycapabilities/tests.capabilities: ----------------------------------------------------------------------
pycapabilities/tests.capabilities: Ran 1 test in 3.810s
pycapabilities/tests.capabilities: 
pycapabilities/tests.capabilities: FAILED (errors=1)
pycapabilities/tests.capabilities: 
pycapabilities/tests.capabilities: 
pycapabilities/tests.capabilities: Exit Code: 1
pycapabilities/tests.capabilities: ################################################################################
pycapabilities/tests.capabilities: Tester failed, reason: CRASH
pycapabilities/tests.capabilities: 
```

This happens when the code tries to import the `capabilities.so` module compile in a previous recipe step. All of this is very similar to what happens with the `hit.so` module, which runs fine :-/

loganharbour commented:
> This happens when the code tries to import the capabilities.so module compile in a previous recipe step. All of this is very similar to what happens with the hit.so module, which runs fine :-/

We've had this issue in the past with a glibc mismatch, can't say I recall the resolution...

dschwen commented:
Wow, this is so messed up... We have two ways of building the `hit.so` python module.

In `moose.mk` we have

```
#
# hit python bindings
#
pyhit_srcfiles  := $(HIT_DIR)/hit.cpp $(HIT_DIR)/lex.cc $(HIT_DIR)/parse.cc $(HIT_DIR)/braceexpr.cc
...
hit $(pyhit_LIB) $(hit_CLI): $(pyhit_srcfiles) $(hit_CLI_srcfiles)
	@echo "Building and linking "$@"..."
	@bash -c '(cd "$(HIT_DIR)" && $(libmesh_CXX) -std=c++17 -w -fPIC -lstdc++ -shared $^ $(pyhit_COMPILEFLAGS) $(DYNAMIC_LOOKUP) -o $(pyhit_LIB))'
	@bash -c '(cd "$(HIT_DIR)" && $(MAKE))'
```

and in `framework/contrib/hit/Makefile` we have

```
hit.so: parse.cc lex.cc braceexpr.cc
	$(CXX) -std=c++17 $(wasp_CXXFLAGS) -w -fPIC -lstdc++ -shared -L$(PYTHONPREFIX)/lib $(PYTHONCFLAGS) $(DYNAMIC_LOOKUP) $^ $(HITCPP) -o $@ $(wasp_LDFLAGS)
```

the first rule calls `make` with the second `Makefile`, but that does not compile the bindings.

dschwen commented:
On rod (or rather apptainer) `/opt/miniforge3/envs/moose/lib/libstdc++.so.6` has the `GLIBCXX_3.4.29` symbols, but

```
[moose-dev-x86_64][~/moose/framework]$ ldd /home/schwd/moose/framework/contrib/capabilities/capabilities.so
/home/schwd/moose/framework/contrib/capabilities/capabilities.so: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /home/schwd/moose/framework/contrib/capabilities/capabilities.so)
	linux-vdso.so.1 (0x00007ffe799fd000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f2948fa1000)
```

the `libstdc++` at `/lib64/libstdc++.so.6` does not.

dschwen commented:
This gets weirder. If I import `pyhit` first then `pycapabilities` imports fine, if I just import `pycapabilities` I get the error

```
[moose-dev-x86_64][~/moose/python]$ /usr/bin/env python3
Python 3.10.13 | packaged by conda-forge | (main, Dec 23 2023, 15:36:39) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyhit
>>> import pycapabilities
>>>
[moose-dev-x86_64][~/moose/python]$ /usr/bin/env python3
Python 3.10.13 | packaged by conda-forge | (main, Dec 23 2023, 15:36:39) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pycapabilities
Building and linking capabilities...
Traceback (most recent call last):
  File "/home/schwd/moose/python/pycapabilities/__init__.py", line 12, in <module>
    import capabilities
ModuleNotFoundError: No module named 'capabilities'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/schwd/moose/python/pycapabilities/__init__.py", line 17, in <module>
    import capabilities
ImportError: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /home/schwd/moose/python/pycapabilities/../../framework/contrib/capabilities/capabilities.so)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/schwd/moose/python/pycapabilities/__init__.py", line 21, in <module>
    import capabilities
ImportError: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.29' not found (required by /home/schwd/moose/python/pycapabilities/../../framework/contrib/capabilities/capabilities.so)
```

dschwen commented:
Fudge. I had removed an `import mooseutils` because it was not used, but it apparently somehow triggers the loading of the correct `libstdc++`. 😵 

```
[moose-dev-x86_64][~/moose/python]$ /usr/bin/env python3
Python 3.10.13 | packaged by conda-forge | (main, Dec 23 2023, 15:36:39) [GCC 12.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import mooseutils
>>> import pycapabilities
>>>
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2113223/) on 7243062 wanted to post the following:

Warning: This PR changes repo size by 5.33 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2113364/) on 47faad5 wanted to post the following:

Warning: This PR changes repo size by 5.14 MiB.

dschwen committed:
Implement PEG parser (#26470)

dschwen commented:
I have a very generic expression parser almost ready. I would allow parenthesis and logic operators. The problem is that it is not possible to output a simple human readable skip reason with that approach (imagine multiple nested parenthesis with and/or and negation operators - which condition leads to the final _false_ state is not easy to back out again).

dschwen committed:
Update test specs (#26470)

dschwen commented:
The tzlaine parser framework now [dropped support for Clang 10](https://github.com/tzlaine/parser/commit/12de83961811c500c31a5c060e254de2bcdfbcc6), which is our current _min clang_. It also produces a ton of compile warnings, some of which cannot seem to be silenced (`-Wtype-limits` kept pooping up despite diagnostic pragmas and - in the latest iteration - a `system_header` pragma. Which should be total overkill).

loganharbour commented:
> The tzlaine parser framework now [dropped support for Clang 10](https://github.com/tzlaine/parser/commit/12de83961811c500c31a5c060e254de2bcdfbcc6)

We can probably bump this to 11 if needed.

dschwen committed:
Add cpp-peglib by Yuji Hirose (#26470)

dschwen committed:
Fix tests improve docs (#26470)

dschwen committed:
Fix rpath issue (#26470)

loganharbour commented on framework/src/base/Capabilities.C:
```diff
@@ -0,0 +1,87 @@
+  auto it_pair = _capability_registry.lower_bound(capability);
+  if (it_pair == _capability_registry.end() || it_pair->first != capability ||
+      it_pair->second.first == value)
+    it_pair = _capability_registry.emplace_hint(it_pair, capability, std::make_pair(value, doc));
```
```suggestion
    _capability_registry.emplace_hint(it_pair, capability, std::make_pair(value, doc));
```

Was confused briefly as to why you used this, looks like you don't

loganharbour commented on framework/src/base/Capabilities.C:
```diff
@@ -0,0 +1,87 @@
+      it_pair->second.first == value)
+    it_pair = _capability_registry.emplace_hint(it_pair, capability, std::make_pair(value, doc));
+  else
+    mooseWarning("Capability '",
```
Should this not be an error instead?

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -634,9 +642,364 @@ MooseApp::MooseApp(InputParameters parameters)
+void
+MooseApp::registerCapabilities()
+{
+  // helper lambdas
+  auto haveCapability = [](const std::string & capability, const std::string & doc)
+  { addCapability(capability, true, doc + " is available."); };
+
+  auto missingCapability =
+      [](const std::string & capability, const std::string & doc, const std::string & help = "")
+  { addCapability(capability, false, doc + " is not available. " + help); };
+
+  auto haveCapabilityVersion =
+      [](const std::string & capability, const std::string & doc, const std::string & version)
+  { addCapability(capability, version, doc + " version " + version + " is available."); };
```
Is it useful to have these as functions instead? that is, are they useful for apps?

dschwen replied:
follow on PR?

loganharbour replied:
fine by me

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -634,9 +642,364 @@ MooseApp::MooseApp(InputParameters parameters)
+
+  {
+    const auto doc = "Workbench Analysis Sequence Processors (WASP) parser library";
+#ifdef WASP_ENABLED
```
I think we can actually get rid of this now, right? WASP needs to always be enabled, we shouldn't be able to even build without it.

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -975,6 +1347,36 @@ MooseApp::setupOptions()
+      }
+      if (status == CapabilityUtils::UNKNOWN)
+      {
+        mooseInfo("Required capabilities '",
```
Should this be an error?

loganharbour commented on modules/solid_mechanics/examples/cframe_iga/tests:
```diff
@@ -7,10 +7,8 @@
-    method = '!DBG'
     # See #24187
-    petsc_version = '>=3.12.0'
+    capabilities = 'petsc>=3.12.0 & !DBG'
```
Not `method!=dbg`?

loganharbour commented on python/TestHarness/testers/RunApp.py:
```diff
@@ -296,7 +301,12 @@ def testExitCodes(self, moose_dir, options, output):
             # in the derived class.
-            if options.valgrind_mode == '' and not specs.isValid('expect_err') and len( [x for x in filter( lambda x: x in output, specs['errors'] )] ) > 0:
+            if self.exit_code == 77:
+                self.setStatus(self.skip, "ABORT")
```
Why `ABORT`? Or is that not even used anywhere, it just shows as skip?

dschwen replied:
I wanted a way to distinguish a test harness initiated static skip and a dynamic skip that runs the app

loganharbour replied:
Maybe `CAPABILITY` or something? `ABORT` is pretty general

loganharbour commented on python/pycapabilities/tests/tests:
```diff
@@ -0,0 +1,9 @@
+  [capabilities]
+    type = PythonUnitTest
+    input = test_capabilities.py
+    requirement = "The system shall ..."
```
...

loganharbour reviewed (dismissed):
Mostly minor.

We need a newsletter entry for changing main and in general referencing the system

loganharbour commented:
It's also worth looking at the framework coverage. There's probably a little more that should be covered.

dschwen commented on framework/doc/content/source/base/Capabilities.md:
```diff
@@ -0,0 +1,35 @@
+The capabilities system API consists of the following static functions in `Moose::Capabilities::`
+
+- `add(name, value)` to register a capability with a given `name` (string) and a `value` of one of the aforementioned types
+- `check(conditions)` to check if the currently registered capability conform to a set of conditions (e.g. `ad_size>50 &method!=dbg & petsc>=3.15.1`)
```
`&` is function parser syntax

dschwen committed:
Address comments (#26470)

dschwen commented:
> It's also worth looking at the framework coverage. There's probably a little more that should be covered.

Link no worky

loganharbour commented:
> > It's also worth looking at the framework coverage. There's probably a little more that should be covered.
> 
> 
> 
> Link no worky

That's cause you're late :) 

Invalidating just the coverage step will rebuild, because all of the state is good and there.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2313676/) on e0ba37c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a419bffe06d72af487b8a3d8bfd685689e9c69ed`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2389419/) on d89b1bb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 90745b3269df9c63e0b9a1c5afc3d7a013d4d138`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2422349/) on 4f49d05 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26471/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 97af19c98a76c47db2b4157c4e73c28f68f45595`

dschwen commented:
I rebased on top of your test harness changes in devel. The main thing I need to sort out is here:
https://github.com/dschwen/moose/blob/b752bb971f88dc540939dce80dcb253eb4cfef5c/python/TestHarness/testers/Tester.py#L419-L433

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2539942/), step Docs: sync website on 8edddd8 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26471/site)

This comment will be updated on new commits.

dschwen commented:
Note to myself:
```
  File "/opt/moose/share/moose/python/pycapabilities/__init__.py", line 18, in <module>
    import capabilities
ImportError: libhit-opt.so.0: cannot open shared object file: No such file or directory
```
I had encountered that before...

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2539989/), step Generate coverage on 8edddd8 wanted to post the following:


## Framework coverage

No coverage report(s) were found for the base commit 019fd714d5.
[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26471/coverage/framework)


## Modules coverage

No coverage report(s) were found for the base commit 019fd714d5.

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26471/coverage/xfem)

</details>



This comment will be updated on new commits.

dschwen committed:
debug rpath issue

loganharbour committed:
Fix merge conflict

loganharbour committed:
Actually test the capability errors

loganharbour committed:
Remove unused code

loganharbour committed:
Register all apps as capabilities

loganharbour committed:
Register optional submodule capabilities and use them

loganharbour committed:
Replace libtorch in test specs

loganharbour committed:
Update more test specs to use capabilities

loganharbour committed:
Setup better error handling for capability parse failures

loganharbour committed:
Remove libtorch requirement for neml2

loganharbour committed:
Remove unreachable code

loganharbour committed:
Register all apps from non-test app in modules

loganharbour committed:
Make an index error a valid failure

loganharbour committed:
Only do the two-tiered check when needed

dschwen committed:
Add newsletter entry (#26470)

dschwen committed:
python tests fix

loganharbour committed:
Convert unique_id to capabilities

loganharbour committed:
Add conversion script for capabilities

loganharbour committed:
Fix duplicate capabilities

loganharbour committed:
Remove old options, use capabilities to determine options

loganharbour committed:
Use version instead of boolean for libtorch capability

loganharbour committed:
Don't use an application for scripts tests

loganharbour committed:
Remove old petsc capability from scripts

loganharbour committed:
Replace config options with capabilities

loganharbour committed:
Add check, only get value if needed

loganharbour committed:
Use capabilities within ifelse extension

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2771583/), step Versioner verify on c88efc5 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 2 failed

| package   | status    | hash    | to hash   | version            | to version         |
|-----------|-----------|---------|-----------|--------------------|--------------------|
| wasp      | NEED BUMP | 46ea1c0 | 943fa32   | 2025.02.25 build 0 | 2025.02.25 build 0 |
| moose-dev | NEED BUMP | ca0a369 | 860f7c1   | 2025.03.06         | 2025.03.06         |

Verification failed.

loganharbour committed:
Remove hasLibtorch and use hasCapability instead

loganharbour committed:
Disable ifelse extension for richards

zachmprince reviewed (changes requested):
I can tell this was a monumental effort. My suggestions are pretty much all semantic. It would be great to make sure that our CI runs the same tests with and without these changes.

loganharbour commented on framework/contrib/capabilities/capabilities.C:
```diff
@@ -0,0 +1,148 @@
+//* This file is part of the MOOSE framework
```
It's separate so that it can be built separately and linked with python (test harness)

loganharbour commented on framework/src/base/MooseApp.C:
```diff
@@ -72,6 +72,10 @@
 #include <sys/utsname.h> // utsname
 #endif
 
+#ifdef LIBTORCH_ENABLED
+#include <torch/version.h>
+#endif
+
```
The reason torch includes are special is because they have stuff that conflicts with the moose namespace. this specific file is really simple - it just has three define macros that describe the version. so no c++ conflicts

dschwen commented on framework/include/utils/CapabilityUtils.h:
```diff
@@ -0,0 +1,50 @@
+ * Return state for check. We use a plain enum because we rely on implicit conversion to int.
+ * Capability checks are run in the test harness using the JSON dump exported from the executable
+ * using `--show-capabilities`. This static check does not take dynamic loading into account,
+ * so some capabilites might not be registered at all. Whenever a test agains an unregistered
+ * capability is made the result is returned as POSSIBLE_FAIL/POSSIBLE_PASS (in case of predicate
```
`!unregistered` is a possible pass if `unregistered` is an unknown capability. `unregistered` is a possible fail. The "possible" here comes from the two options of 1) assuming an unregistered capability is simply false, or 2) that it might be registered at runtime through dynamic loading. If we forego option 2 we just remove the "possible" and get the correct result. Otherwise, if we encounter any of the uncertain results we can dispatch a runtime capability check, which would then take dynamically registered capabilities into account.

loganharbour replied:
Changing this to:

```
/**
 * Return state for check. We use a plain enum because we rely on implicit conversion to int.
 * Capability checks are run in the test harness using the JSON dump exported from the executable
 * using `--show-capabilities`. This static check does not take dynamic loading into account,
 * as capabilities that would be registered after initializing the dynamically loaded application
 * will not exist with `--show-capabilities`.
 *
 * A requested capability that is not registered at all is considered in a "possible" state,
 * as we cannot guarantee that it does or not exist with a dynamic application. If no dynamic
 * application loading is used, the possible states can be considered certain states.
 *
 * When the test harness Tester specification "dynamic_capabilities" is set to True,
 * it will run the test unless the result of the check is CERTAIN_FAIL. In this case,
 * the runtime check in the executable will terminate if the result is either CERTAIN_FAIL
 * or POSSIBLE_FAIL.
 */
```

zachmprince replied:
beautiful

dschwen commented on framework/src/utils/CapabilityUtils.C:
```diff
@@ -0,0 +1,322 @@
+Result
+check(std::string requirements, const Registry & app_capabilities)
+{
+  using namespace peg;
```
I would argue that since this is scoped to a single function we should leave it as it is...

loganharbour replied:
Yeah, I would prefer to not do this as well.

loganharbour commented on modules/fluid_properties/test/tests/external_submodules/tests:
```diff
@@ -11,47 +11,47 @@
       threading = '!pthreads'
       detail = 'spline-based table-lookup air fluid properties,'
-      required_submodule = 'contrib/air'
+      capabilities = 'airapp'
```
I could remove this. I dunno, what do you think @dschwen

dschwen replied:
I like the app suffix as disambiguation.

loganharbour replied:
Yeah. We use the real app type in inputs for multiapps. This should probably just be consistent.

dschwen commented on framework/src/utils/CapabilityUtils.C:
```diff
@@ -0,0 +1,322 @@
+  if (requirements.length() == 0)
+    return {CapabilityUtils::CERTAIN_PASS, "Empty requirements", ""};
+
+  parser parser(R"(
+    Expression    <-  _ Bool _ LogicOperator _ Expression / Bool _
+    Bool          <-  Comparison / '!' Bool / '!' Identifier / Identifier / '(' _ Expression _ ')'
+    Comparison    <-  Identifier _ Operator _ Version / Identifier _ Operator _ String
+    String        <-  [a-zA-Z0-9_-]+
+    Identifier    <-  [a-zA-Z][a-zA-Z0-9_]*
+    Operator      <-  [<>=!]+
+    LogicOperator <-  [&|]
+    Version       <-  Number '.' Version  / Number
+    Number        <-  [0-9]+
+    ~_            <-  [ \t]*
+  )");
+
```
Probably not constexpr. If we make this static we'd need to wrap it into a class and put all the callback registrations into the constructor. We should profile this first though, and I'd be happy to do this in a follow on PR.

loganharbour replied:
yeah, I'd like to defer this.

dschwen commented on framework/src/utils/CapabilityUtils.C:
```diff
@@ -0,0 +1,322 @@
+  while (true)
+  {
+    const auto len = requirements.length();
+    if (len >= 2 && ((requirements[0] == '\'' && requirements[len - 1] == '\'') ||
+                     (requirements[0] == '"' && requirements[len - 1] == '"')))
```
the `\` is a `\'` ... that _is_ the single quote

dschwen commented on modules/combined/test/tests/optimization/invOpt_elasticity_modular/tests:
```diff
@@ -11,8 +11,7 @@
     max_threads = 1
     # steady solve
     recover = false
-    libtorch = true
```
Yes, neml2 depends on libtorch

dschwen commented on modules/contact/test/tests/verification/patch_tests/plane_3/tests:
```diff
@@ -10,8 +10,7 @@
     mesh_mode = REPLICATED
-    superlu = true
-    petsc_version = '>=3.7.6'
+    capabilities = 'superlu'
```
Yeah, good question. At some point I just don't see the point of checking for versions this old anymore. We're not testing against them, so we don;t even know for sure that MOOSE still works properly with those old versions. Along the same lines, we should probably decide on a minimal PETSc version that we guarantee to support, and get rid of any older ifdef'ed code.

loganharbour replied:
I agree. We actually use to have an alternate petsc, but we removed that testing. At this point, we only really qualify with the _current_ version of the petsc submodule at that time so I don't really see this being an issue

dschwen commented on python/TestHarness/TestHarness.py:
```diff
@@ -292,81 +298,88 @@ def __init__(self, argv: list, moose_dir: str, moose_python: str, test_root: Tes
+            threading = None
+            threads = get_option('threads', from_type=bool)
+            if threads:
+                for name in ['tbb', 'openmp']:
+                    if get_option(name, from_type=bool):
+                        threading = name
+                        break
+                if not threading:
+                    threading = 'pthreads'
```
beautiful

loganharbour commented on modules/navier_stokes/test/tests/finite_volume/two_phase/mixture_model/tests:
```diff
@@ -35,7 +35,7 @@
     requirement = 'The system shall be able to solve two-phase flow in a channel using the mixture drift flux model.'
     installation_type = in_tree
-    min_ad_size = 80
+    capabilities = 'ad_size>=80'
```
Yeah, I realized this as well. This hasn't broken anything (we usually just estimate this value anyway), so I'm content with using the named behavior instead of the actual behavior

loganharbour commented on modules/thermal_hydraulics/test/tests/components/hs_coupler_2d3d/tests:
```diff
@@ -86,7 +86,7 @@
       cli_args = 'Components/hs3d/file=mesh/mesh_fine.e'
       expect_err = 'MOOSE must be configured with a larger AD container size'
-      max_ad_size = 55
+      capabilities = 'ad_size<=55'
```
I agree; same as https://github.com/idaholab/moose/pull/26471#discussion_r2013262810

loganharbour commented on python/TestHarness/TestHarness.py:
```diff
@@ -985,7 +1000,29 @@ def writeResults(self, complete=False, stats=None):
+        if exe_path:
+            return exe_path
+
+        print(f'ERROR: Failed to find the executable {name}')
+        sys.exit(1)
```
This would be a pretty common error. Imagine calling run tests without a build.

Why see an exception? The error is pretty clear here

zachmprince replied:
An exception could be ignored with a `try/except`. Would this not cause issues if you are just doing python testing without a moose executable?

loganharbour replied:
Agreed with @zachmprince on this one - had `getExecutable()` throw, and caught that specific exception in the one use case where we run `getExecutable()` early. He gave the :+1:

loganharbour commented on python/TestHarness/schedulers/Job.py:
```diff
@@ -306,7 +306,14 @@ def getCommandRan(self):
-        return self.__tester.getRunnable(self.options)
+        try:
+            runnable = self.__tester.getRunnable(self.options)
+        except:
+            trace = traceback.format_exc()
+            self.setStatus(self.error, 'TESTER EXCEPTION')
+            self.appendOutput(util.outputHeader('Python exception encountered') + trace)
+            return False
```
This happens in a thread. The main thread controls the output, and it is captured (for example)

```
materials/derivative_material_interface.warn: Working Directory: /data/harblh/projects/moose/test/tests/materials/derivative_material_interface
materials/derivative_material_interface.warn: Running command: /data/harblh/projects/moose/test/moose_test-opt -i warn.i --error --error-override --libtorch-device cpu
materials/derivative_material_interface.warn: 
materials/derivative_material_interface.warn: ################################################################################
materials/derivative_material_interface.warn: Begin job output
materials/derivative_material_interface.warn: 
materials/derivative_material_interface.warn: ################################################################################
materials/derivative_material_interface.warn: Python exception encountered
materials/derivative_material_interface.warn: ################################################################################
materials/derivative_material_interface.warn: Traceback (most recent call last):
materials/derivative_material_interface.warn:   File "/data/harblh/projects/moose/python/TestHarness/schedulers/Job.py", line 310, in getRunnable
materials/derivative_material_interface.warn:     runnable = self.__tester.getRunnable(self.options)
materials/derivative_material_interface.warn:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
materials/derivative_material_interface.warn:   File "/data/harblh/projects/moose/python/TestHarness/testers/Tester.py", line 310, in getRunnable
materials/derivative_material_interface.warn:     raise Exception('This is bad')
materials/derivative_material_interface.warn: Exception: This is bad
materials/derivative_material_interface.warn: ################################################################################
materials/derivative_material_interface.warn: Python exception encountered
materials/derivative_material_interface.warn: ################################################################################
materials/derivative_material_interface.warn: Traceback (most recent call last):
materials/derivative_material_interface.warn:   File "/data/harblh/projects/moose/python/TestHarness/schedulers/Job.py", line 310, in getRunnable
materials/derivative_material_interface.warn:     runnable = self.__tester.getRunnable(self.options)
materials/derivative_material_interface.warn:                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
materials/derivative_material_interface.warn:   File "/data/harblh/projects/moose/python/TestHarness/testers/Tester.py", line 310, in getRunnable
materials/derivative_material_interface.warn:     raise Exception('This is bad')
materials/derivative_material_interface.warn: Exception: This is bad
materials/derivative_material_interface.warn: 
materials/derivative_material_interface.warn: ################################################################################
materials/derivative_material_interface.warn: End job output
materials/derivative_material_interface.warn: 
materials/derivative_material_interface.warn .................... FAILED (TESTER EXCEPTION)
```

loganharbour commented on python/TestHarness/util.py:
```diff
@@ -607,14 +515,66 @@ def checkLibtorchVersion(checks, test):
+    [status, message, doc] = pycapabilities.check(requested, supported)
+    if status == pycapabilities.PARSE_FAIL:
+        raise ValueError(f"Failed to parse capabilities='{requested}'")
+    success = status == pycapabilities.CERTAIN_PASS or (status == pycapabilities.POSSIBLE_PASS and not certain)
```
It has to be CERTAIN_PASS when we do not allow dynamic apps (`dynamic_capabilities = False` in the spec, the default). When `dynamic_capabilities = True`, then `certain=False` here and it could be `POSSIBLE_PASS`

loganharbour commented on python/peacock/tests/input_tab/InputTree/tests:
```diff
@@ -97,6 +97,6 @@
       detail = "handles incompatible parameter changes without failure."
     []
-    platform = LINUX # https://github.com/conda-forge/vtk-feedstock/issues/109
+    capabilities = 'platform=linux' # https://github.com/conda-forge/vtk-feedstock/issues/109
```
A moose app isn't available for the gui, scripts, and python tests. Which means capabilities don't work.

And... this wasn't caught because we don't test the gui anymore (it's no longer supported)

loganharbour commented on python/pycapabilities/tests/test_capabilities.py:
```diff
@@ -0,0 +1,34 @@
+class TestCapabilities(unittest.TestCase):
+    def test(self):
+
+        # Load the packages
+        import pycapabilities
```
I think that this just needs to be late due to the potential build of the capabilities library. @dschwen ?

loganharbour replied:
Confirmed with Daniel. This just needs to be done late.

loganharbour commented on scripts/convert_to_capabilities.py:
```diff
@@ -0,0 +1,214 @@
+
+MOOSE_DIR = os.environ.get('MOOSE_DIR', os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..')))
+sys.path.append(os.path.join(MOOSE_DIR, 'python'))
+import pyhit
```
It needs to be after 7 and 8 due to the path addition

loganharbour commented on scripts/convert_to_capabilities.py:
```diff
@@ -0,0 +1,214 @@
+      node.removeParam(param)
+
+    # min_ad_size
+    do_min_ad_size = lambda value: f'ad_size>={value}'
+    fix_param('min_ad_size', do_min_ad_size)
+    # max_ad_size
+    do_max_ad_size = lambda value: f'ad_size<={value}'
+    fix_param('max_ad_size', do_max_ad_size)
```
While this matches current behavior, I'd still prefer to match the named behavior

loganharbour commented on framework/src/utils/CapabilityUtils.C:
```diff
@@ -0,0 +1,322 @@
+{
+
+Result
+check(std::string requirements, const Registry & app_capabilities)
```
```c++
/**
 * Checks if a set of requirements is satisified by the given capability registry
 *
 * @param requirements The requirement string
 * @param capabilities The registry that contains the capabilities
 *
 * This method is exposed to Python within the capabilities_check method in
 * framework/contrib/capabilities/capabilities.C. This external method is used
 * significantly by the TestHarness to check capabilities for individual test specs.
 *
 * Additionally, this method is used by the MooseApp command line option
 * "--required-capabilities ...".
 *
 * Requirements can use comparison operators (>,<,>=,<=,=!,=), where the name of
 * the capability must always be on the left hand side. Comparisons can be performed
 * on strings "compiler!=GCC" (which are case insensitive), integer numbers
 * "ad_size>=50", and version numbers "petsc>3.8.0". The state of a boolean
 * valued capability can be tested by just specifying the capability name "chaco".
 * This check can be inverted using the ! operator as "!chaco".
 *
 * The logic operators & and | can be used to chain multiple checks as
 * "thermochimica & thermochimica>1.0". Parenthesis can be used to build
 * complex logic expressions.
 *
 * See the description for CheckState for more information on why a
 * certain state would be returned.
 */
Result check(std::string requirements, const Registry & capabilities);
```

Added more context

loganharbour committed:
Address review: python cleanup, add more documentation

loganharbour committed:
Use an exception instead when the app is not found

loganharbour committed:
Add excerpt on how to add a capability to an app

loganharbour commented on framework/doc/content/source/base/Capabilities.md:
```diff
@@ -0,0 +1,35 @@
+
+The capabilities system API consists of the following static functions in `Moose::Capabilities::`
+
+- `add(name, value)` to register a capability with a given `name` (string) and a `value` of one of the aforementioned types
+- `check(conditions)` to check if the currently registered capability conform to a set of conditions (e.g. `ad_size>50 & method!=dbg & petsc>=3.15.1`)
+- `dump()` to build a string with a JSON dictionary of all currently registered capabilities
```
Added an excerpt on the more important part of how to actually register a capability within an app (which should be the only part of this used by a standard developer) in https://github.com/idaholab/moose/pull/26471/commits/604ee411481b0ae64e513c203bcd31224cd7420f

loganharbour committed:
Add newsletter entry describing convert_to_capabilities script

dschwen committed:
Depend on revision header

loganharbour committed:
Re-add hasLibtorch method

Commit:
Fix module activation routine

This was barely working before because of the friendly symlinks we
were creating in CONDA_PREFIX/bin. As soon as we removed those
symlinks, our really bad module activation routine surfaces.

loganharbour committed:
Attempt to fix capabilities.so linking on mac installs

loganharbour committed:
Re-run capability conversion

loganharbour commented:
Python and VTB failures are unrelated. VTB patch is in.

loganharbour commented:
The capabilities extension in moosedocs is used to show or hide content based on what capabilities an application has.

In the case of the VTB - it doesn't have a single application. The use of capabilities for it is not appropriate. The solution above is likely to disable the ifelse extension.
